[x11] New window stacking logic

elParaguayo opened the pull request:
Draft for now!

I've hated the window stacking logic in our x11 backend for a while. Especially because I wrote a lot of it.

Having looked at the new Wayland backend and seen the concept of sub-trees for different layer groups I felt that was a cleaner way to manager stacking in general. So this PR is an attempt to implement that sort of stacking.

The key idea is a new class `ZManager` which maintains a snapshot of the stacking order of windows across different layer groups. When a window requests to be moved in the stack, the manager updates the stack and then tells the window which other window it needs to be stacked above/below and the window then sends this request to the X server.

This will break a lot of things:
* ~It presumably breaks wayland entirely as there's no implementation of `ZManager` in the wayland backend. Will sort soon.~
* ~I need some logic to restack windows when they lose focus (e.g. fullscreen windows are elevated when they're focused so we need to know when they lose focus. Hooks?)~
* More stuff that I haven't figured out yet.

Also:
* Needs more tests
* Needs more documentation

elParaguayo committed:
Add z layer manager

elParaguayo committed:
More layer manager code

elParaguayo committed:
Adding to x11

elParaguayo committed:
Adding to x11 contd.

elParaguayo committed:
Make zmanager abstract

elParaguayo committed:
Tidying up.

elParaguayo committed:
Updates

elParaguayo committed:
Fix last test

tych0 commented on libqtile/backend/base/zmanager.py:
```diff
@@ -0,0 +1,113 @@
+    BRING_TO_FRONT = 7
+    TOP = 8
+    OVERLAY = 9
+    SYSTEM = 10
```
Do you need something here that's OVERRIDE_REDIRECT i.e. "don't mess with the stacking of this window"? But I guess it's not really "don't mess with it" but "keep it relative to its parent window"?

elParaguayo committed:
Fix more tests

elParaguayo committed:
move stack to zmanager

elParaguayo committed:
better handling of float state

elParaguayo committed:
Restack unfocused windows

elParaguayo commented:
I'm going to remove `ZManager` from `backend.base` and `backend.wayland` as Wayland just doesn't need this. While it could be made to work, it would just introduce unnecessary steps. `ZManager` should be invisible to users and they'd just access the usual API calls `move_up` etc etc

elParaguayo committed:
Remove zmanager from wayland backend

elParaguayo committed:
Fixing pre-commit check errors

elParaguayo committed:
Remove wayc-related code

elParaguayo commented:
NB for anyone testing this PR, I've had to rebase and force-push the changes in order to get CI to run correctly.

elParaguayo committed:
ZManager - stacking tree

elParaguayo committed:
Handle scratchpads

elParaguayo committed:
force bring_to_front call to top

MoonCubes committed:
Dont't add internal windows to _NET_CLIENT_LIST_STACKING

This commit only adds windows that are in a group to _NET_CLIENT_LIST_STACKING.
This fixes internal windows showing in window switching programs
like rofi.

elParaguayo committed:
Fix tests and change layer on focus change

elParaguayo committed:
Fix the 'fix'

elParaguayo committed:
Fix wayland leftovers

elParaguayo committed:
Internal windows in their own layer?

elParaguayo committed:
node info

elParaguayo commented:
@nazar-pc I've rebased this.

tych0 commented on libqtile/backend/base/stacking.py:
```diff
@@ -0,0 +1,41 @@
+from enum import IntEnum, auto
+
+
+class LayerGroup(IntEnum):
```
Does this need to be in `backend.base` given that it's all X11 specific?

tych0 commented on libqtile/backend/x11/core.py:
```diff
@@ -278,10 +290,10 @@ def on_config_load(self, initial) -> None:
 
             self.update_client_lists()
-            win.change_layer()
+            # win.change_layer()
```
You can drop this.

tych0 commented on libqtile/backend/x11/core.py:
```diff
@@ -449,27 +461,7 @@ def display_name(self) -> str:
-                stacked_wids.append(wid)
-
-        self._root.set_property("_NET_CLIENT_LIST_STACKING", stacked_wids)
+        self.zmanager.update_client_lists()
```
Can the X11 Core inherit from Zmanager instead of having this proxy / reference loop between zmanager and core?

elParaguayo replied:
Yes. That's a nice idea.

tych0 reviewed (commented):
For what it's worth, I have been running this branch for a while and haven't seen anything weird.

elParaguayo commented:
> For what it's worth, I have been running this branch for a while and haven't seen anything weird.

Pleased to hear it. There are some little issues that we're working through in #4805 but I think we're close to something that can be merged and then we just submit patches for any remaining issues.

elParaguayo committed:
Tych0 changes

elParaguayo committed:
Move stacking info method

elParaguayo commented:
I assume moving the parent to the top if a child opens isn't what you want either. If that's the case, are we just going to ignore the parent/child relationship for stacking purposes? 

The end goal is to have something workable so we can update this as needed.

elParaguayo commented:
Thanks. At the moment, transient windows are just displayed immediately above their parent but the feedback I've received is consistent with yours: it's not a good approach.

3hhh commented:
The last 3 months of my testing have not brought up any further issues, i.e. I believe this one just needs the `transient_for` fix and is then good to land into `master`.

elParaguayo committed:
Restore missing test

elParaguayo commented:
OK - I've rebased this - we need to get this landed. @richcarni we should be mirroring behaviour in wayland - how are transient windows handled there?

richcarni commented:
> we should be mirroring behaviour in wayland - how are transient windows handled there?

In x11 are transients essentially just un-managed children? The wayland equivalent of popups - menus and simple dialogs? I ask because I'm not sure if other child windows, like a save dialogs are considered transients

Anyway, assuming we're talking about un-manged children, they are currently just added to the top of the parent's layer. I was playing around with putting them under the parent node in #5815. I think I am of the same opinion as you that transients should be fixed to (and above) their parent. I know technically that means it's possible if you're working in a window that is below another, it's transient could be hidden, but to me, this seems like it should be an unusual scenario. In most cases I imagine that the window you are working with should be on top. If there is demand to maintain the existing behaviour, that could be a config option?

I haven't had a good look at this PR for a while so I'll try and do that. Bring to front behaviour is one that I would like wayland to mirror. I'm also curious whether you're currently using MAX and FULLSCREEN layers

nazar-pc commented:
I did some digging and found that I'm still running commit 5f4a4c9eb7401b5cc93cf9da6b6b4933e75feda0 for almost a year now, that is the best version of this PR that I have tried. It is part of the tip 7739244b90a57f14753459260a783a64b0e162d7 before the first force-push into this PR, that tip is possibly fine too, though I have not tried it.

And then I cherry-picked https://github.com/qtile/qtile/pull/5670 on top.

If you could ship a version of 5f4a4c9eb7401b5cc93cf9da6b6b4933e75feda0, I would be mostly (though not perfectly) happy. Current upstream state is too frustrating to use, I'm not sure how you folks are doing it.

UPD: You can access it like this:
```bash
git fetch upstream 7739244b90a57f14753459260a783a64b0e162d7
git checkout 7739244b90a57f14753459260a783a64b0e162d7
```

richcarni commented:
@nazar-pc Are the issues you're experiencing mostly of the windows not appearing above other windows variety?

Possibly somewhat related, I have a new idea for keep above behaviour on wayland. I'm thinking of giving every layer it's own keep_above layer - so transients/floats kept above should work regardless of whether the window is tiling, maximized, fullscreen, scratchpad. (I've noticed an issue with scratchpad dialogs not showing) 

Also I think transients should maintain keyboard focus until closed, even with follow_mouse_focus on. It's very easy for a dialog to lose focus because the mouse may have accidentally moved 1 pixel

nazar-pc commented:
@richcarni yes, see https://github.com/qtile/qtile/issues/4805 for a lengthy discussion (it'd be ideal to be familiar with it before commenting there further).

It still happens in certain cases even with the revision I'm using and about once a month something glitches out and I have to restart Qtile to fix relative positions of windows, but other that that it works reasonably well.

`keep_above` sounds interesting and like it might fix issues with the current state of this PR. Though I'm not sure what is the point of it compared to simply having each window in its own layer.

The hierarchy used here seemed nice at first, but ended up frustrating to use in practice.

elParaguayo committed:
How about this?

elParaguayo commented:
Based on discussion in #5815 I've implemented these rules:

- When a parent spawns a child, the child is moved to the top of the layer group containing the parent
- We store a reference between parents and children
- If a child is moved down, it cannot be moved below its parent - if this would happen then the parent must be moved down too
- Similarly, a parent should not be above its child - if that would happen then the child must be moved up too

Almost certainly left some bugs in there but would be good to get any feedback.

Pinging @richcarni in case you want to test this. This is all in python so would be easy to transfer to wayland too.

nazar-pc commented:
I just tried 6865969f0f2ae6f3ad927632343e4a32f4c8aaea and it doesn't work correctly with floating windows.

Any floating window (like Ulauncher or Nextcloud settings) is covered by any non-floating window that started later.

elParaguayo commented:
Oof. Sorry, that's clearly wrong. Let me see what's going on

richcarni commented:
> Pinging @richcarni in case you want to test this. This is all in python so would be easy to transfer to wayland too.

Good to see you're getting some good feedback from an x11 user!

If it's not hard to transfer to wayland I'd be happy to start testing that. Or even helping out do that

elParaguayo commented:
Yes. It's easy enough. I'm try and take a look over the weekend

elParaguayo committed:
Fix stacking info

elParaguayo commented:
> I just tried [6865969](https://github.com/qtile/qtile/commit/6865969f0f2ae6f3ad927632343e4a32f4c8aaea) and it doesn't work correctly with floating windows.
> 
> Any floating window (like Ulauncher or Nextcloud settings) is covered by any non-floating window that started later.

@nazar-pc  I can't even get `ulauncher` working for me to test this. Can you do an `xprop` on the floating window please and also run `qtile cmd-obj -o core -f stacking_info` (you'll need latest commit to run that).

nazar-pc commented:
<details>
<summary>With Ulauncher window not visible</summary>

```json
{
  "name": null,
  "x": 0,
  "y": 0,
  "type": "Tree root",
  "wid": 0,
  "children": [
    {
      "name": "BACKGROUND",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "BOTTOM",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "KEEP_BELOW",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "INTERNAL",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "LAYOUT",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": [
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        },
        {
          "name": "Home",
          "x": 3840,
          "y": 1728,
          "type": "Window",
          "wid": 25165867,
          "children": []
        },
        {
          "name": "qtile cmd-obj -o cor ~",
          "x": 3840,
          "y": 1728,
          "type": "Window",
          "wid": 20971523,
          "children": []
        },
        {
          "name": "[x11] New window stacking logic by elParaguayo \u00b7 Pull Request #5397 \u00b7 qtile/qtile \u2014 Firefox Nightly",
          "x": 1536,
          "y": 1728,
          "type": "Window",
          "wid": 25165827,
          "children": []
        },
        {
          "name": "Nextcloud",
          "x": 1536,
          "y": 3024,
          "type": "Window",
          "wid": 8388625,
          "children": []
        }
      ]
    },
    {
      "name": "KEEP_ABOVE",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": [
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        },
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        },
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        },
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        }
      ]
    },
    {
      "name": "MAX",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "FULLSCREEN",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "BRING_TO_FRONT",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "TOP",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "OVERLAY",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "SYSTEM",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    }
  ]
}
```

</details>

<details>
<summary>With Ulauncher window visible</summary>

```json
{
  "name": null,
  "x": 0,
  "y": 0,
  "type": "Tree root",
  "wid": 0,
  "children": [
    {
      "name": "BACKGROUND",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "BOTTOM",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "KEEP_BELOW",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "INTERNAL",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "LAYOUT",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": [
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        },
        {
          "name": "Home",
          "x": 3840,
          "y": 1728,
          "type": "Window",
          "wid": 25165867,
          "children": []
        },
        {
          "name": " sleep 1 && qtile cm ~",
          "x": 3840,
          "y": 1728,
          "type": "Window",
          "wid": 20971523,
          "children": []
        },
        {
          "name": "[x11] New window stacking logic by elParaguayo \u00b7 Pull Request #5397 \u00b7 qtile/qtile \u2014 Firefox Nightly",
          "x": 1536,
          "y": 1728,
          "type": "Window",
          "wid": 25165827,
          "children": []
        },
        {
          "name": "Nextcloud",
          "x": 1536,
          "y": 3024,
          "type": "Window",
          "wid": 8388625,
          "children": []
        }
      ]
    },
    {
      "name": "KEEP_ABOVE",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": [
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        },
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        },
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        },
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        }
      ]
    },
    {
      "name": "MAX",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "FULLSCREEN",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "BRING_TO_FRONT",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": [
        {
          "name": "Ulauncher - Application Launcher",
          "x": 3537,
          "y": 864,
          "type": "Window",
          "wid": 10485767,
          "children": []
        }
      ]
    },
    {
      "name": "TOP",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "OVERLAY",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    },
    {
      "name": "SYSTEM",
      "x": 0,
      "y": 0,
      "type": "Layer group",
      "wid": 0,
      "children": []
    }
  ]
}
```

</details>

But it is not specific to Ulauncher, it is basically anything of that sort. For example, Tint2 panel is supposed to be above everything on every desktop, but it is easily covered by terminal window that launched after it.

Firefox's "Open" dialog is initially in the front of the terminal window that started after Firefox, but as soon as I focus terminal window, "Open" dialog hides behind it and only comes to the front with `lazy.window.bring_to_front()`, but focusing on terminal window brings it back behind again.

elParaguayo commented:
OK - there's something very weird happening there as the same ulauncher window appears multiple times in the stacking tree.

richcarni commented:
@elParaguayo what do you think about creating a dedicated branch for this? I'd like to chip in - maybe work on the wayland side

elParaguayo commented:
I think that's a good idea. Let me rebase and split out the backend-agnostic code into a base module. Will try to do today but life's busy at the moment.
